src/bp_tree_children.erl

```diff
@@ -541,4 +496,130 @@ get_pos(It, Pos, CurrentPos, _TmpAns) ->
+                    Tree3 = gb_trees:enter(ExistingKey, NextValue, Tree2),
+                    {ok, Children#bp_tree_children{data = Tree3}, [Key]};
+                _ ->
+                    Tree2 = gb_trees:insert(Key, Value0, Tree),
+                    {Tree3, AddedKeys} = case LV of
+                        ?NIL ->
+                            add_until(Tree2, MaxSize, NextItems, [Key], ?NIL);
+                        _ ->
+                            {LargestKey, _} = gb_trees:largest(Tree),
+                            add_until(Tree2, MaxSize, NextItems, [Key], LargestKey)
+                    end,
+                    {ok, Children#bp_tree_children{data = Tree3}, AddedKeys}
+            end;
+        none ->
+            case Selector of
+                both ->
+                    {Value, NextValue} = Value0,
+                    Tree2 = gb_trees:insert(Key, Value, Tree),
+                    {ok, Children#bp_tree_children{data = Tree2,
+                        last_value = NextValue}, [Key]};
+                _ ->
+                    Tree2 = gb_trees:insert(Key, Value0, Tree),
+                    {Tree3, AddedKeys} = case LV of
+                        ?NIL -> add_until(Tree2, MaxSize, NextItems, [Key], ?NIL);
+                        _ -> {Tree2, [Key]}
+                    end,
+                    {ok, Children#bp_tree_children{data = Tree3}, AddedKeys}
+            end
+    end.
+
+%%--------------------------------------------------------------------
+%% @private
+%% @doc
+%% Adds {Key, Value} pairs to tree until tree is full or Key is greater then
+%% last key in the tree.
+%% @end
+%%--------------------------------------------------------------------
+-spec add_until(gb_trees:tree(), non_neg_integer() | undefined,
+    [{key(), value()}], [key()], key()) -> {gb_trees:tree(), [key()]}.
+add_until(Tree, MaxSize, [{NextKey, NextValue} | NextTail], AddedKeys,
+    LastKey) when NextKey < LastKey ; LastKey =:= ?NIL ->
+    case (MaxSize =:= undefined) orelse (gb_trees:size(Tree) < MaxSize) of
+        true ->
+            try
+                Tree2 = gb_trees:insert(NextKey, NextValue, Tree),
+                add_until(Tree2, MaxSize, NextTail, [NextKey | AddedKeys], LastKey)
+            catch
+                _:_ ->
+                    % Already exists
+                    {Tree, AddedKeys}
+            end;
+        _ ->
+            {Tree, AddedKeys}
+    end;
+add_until(Tree, _MaxSize, _, AddedKeys, _) ->
+    {Tree, AddedKeys}.
+
+%%--------------------------------------------------------------------
+%% @private
+%% @doc
+%% Removes a keys and associated values.
+%% @end
+%%--------------------------------------------------------------------
+-spec remove({selector(), key()}, remove_pred(), children(), gb_trees:iter(),
+    [{key(), remove_pred()}]) -> {ok, children(), [key()]} | {error, term()}.
+remove({Selector, Key}, Pred, #bp_tree_children{data = Tree} = Children, It, NextItems) ->
```

Comment:
nie dałoby się tej funkcji jakoś rozbić?
Nie lubię jak w jednek funkcji mamy aż  5 zagnieżdzonych case'ów źle się to czyta